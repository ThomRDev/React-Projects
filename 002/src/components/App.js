import logo from '../logo.svg';
import './../styles/App.css';

import PropTypes from "prop-types"

import {useState} from "react"

const  App = ({ initValue }) => {

  const [counter,setCounter] = useState(initValue)

  console.log("renderizo");
  const handleAdd = (event) =>{
    // setCounter(counter + 1)
    setCounter(oldValue=>oldValue+1)
  }
  const handleSubtract = (event) =>{
    if(counter <= 0) return
    // setCounter(counter - 1)
    setCounter(oldValue=>oldValue-1)
  }

  const handleReset = (event)=>{
    setCounter(0)
  }
  return (
    <div className="App">
      <header className="App-header">
      { "renderizo normal"+(new Date().getTime()) }
        <img src={logo} className="App-logo" alt="logo" />
        <div className="value">{ counter }</div>
        <div className='actions'>
          <button onClick={handleAdd}>+1</button>
          <button onClick={handleReset}>Reset</button>
          <button onClick={handleSubtract}>-1</button>
        </div>
      </header>
    </div>
  );
}

App.propTypes = {
  initValue : PropTypes.number
}
App.defaultProps = {
  initValue : 1
}

export default App;
