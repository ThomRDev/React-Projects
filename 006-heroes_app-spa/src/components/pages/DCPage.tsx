import HeroList from '../sections/HeroList'

const DCPage = () => {
  return (
    <div>
      <h1> MarvelScreen </h1>
      <HeroList publisher='DC Comics' />
    </div>
  )
}

export default DCPage