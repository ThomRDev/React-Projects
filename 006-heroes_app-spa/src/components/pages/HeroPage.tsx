import { useMemo } from "react"
import { Navigate , useNavigate, useParams } from "react-router-dom"
import { getHeroById } from "../../selectors/getHeroById"

const HeroPage = () => {


  const { heroId } = useParams()

  const hero = useMemo(()=>getHeroById(heroId as string),[heroId])
  
  // tambien podriamos usar el hook useNavigate
  // pero tendria que teminar en un return vacio
  // no cual ya no seria en Functional Component
  if(!hero) return <Navigate to="/" />
  
  const history = useNavigate()

  const { id, superhero, first_appearance, characters, alter_ego, publisher } = hero;

  // si no colocamos la configuracion de vite.config.ts
  // llamariamos asi /assets/heroes/${id}.jpg
  const imgPath = `/heroes/${id}.jpg`;
  
  const hanldeReturn = () => {
    // history(-1) que valla al historial anterior, pagina anterior
    history(-1)
  }
  
  return (
    <div className="row mt-5">
      <div className="col-4">
        <img
          src={imgPath}
          alt={id}
          className="img-thumbnail animate__animated animate__fadeInLeft"
        />
      </div>

      <div className="col-8">
        <h3>{superhero} </h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Alter Ego:</b> {alter_ego}
            <br />
            <b>Publisher:</b> {publisher}
            <br />  
            <b>First Appearance:</b> {first_appearance}
          </li>
        </ul>

        <h5 className="mt-3">Characters</h5>
        <p>{characters}</p>

        <button className="btn btn-outline-info"
          onClick={hanldeReturn}
        >
          Go back...
        </button>
      </div>
    </div>
  );
}

export default HeroPage