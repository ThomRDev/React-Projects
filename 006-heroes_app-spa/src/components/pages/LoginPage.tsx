import { useContext } from "react"
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../../contexts/AuthContext"
import { ActionTypes } from "../../types/actions"

const LoginPage = () => {
  const history = useNavigate();
  const { dispatch } = useContext(AuthContext)

  return (
    <div>
      <h1>LoginPage</h1>
      <button className="btn btn-outline-primary" onClick={()=>{

        // podria abstraer esto en una funcion login q recibe como parametro al usuario
        // tambien podria invocar al localStorage y almacenar al usuario pero ya lo hago en el provider dentro de un useEffect
        // localStorage.setItem('user', JSON.stringify(user));
        dispatch?.({
          type:ActionTypes.LOGIN,
          payload:{
            name : "ThomR"
          }
        })


        const lastPath = localStorage.getItem("lastPath") ?? "/"
        console.log(lastPath)
        // replace, para evitar regresar a la ruta anterior con el boton de back del navegador
        // que reemplace la historia en lugar de almacenr
        history(lastPath,{
          replace:true
        })
      }} >Login</button>
    </div>
  )
}

export default LoginPage