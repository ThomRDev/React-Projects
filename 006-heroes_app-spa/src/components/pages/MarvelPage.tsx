import React from 'react'
import HeroList from '../sections/HeroList'

const MarvelPage = () => {
  return (
    <div>
      <h1> MarvelScreen </h1>
      <HeroList publisher='Marvel Comics' />
    </div>
  )
}

export default MarvelPage