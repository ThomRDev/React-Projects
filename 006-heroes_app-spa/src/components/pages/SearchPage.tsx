import queryString from "query-string"
import { FormEvent, useMemo, useRef } from "react"
import { useLocation, useNavigate } from "react-router-dom"
import { useForm } from "../../hooks/useForm"
import { getHerosByName } from "../../selectors/getHeroesByName"
import HeroCard from "../cards/HeroCard"

const SearchPage = () => {

  const history = useNavigate()
  const location = useLocation()

  // query-string nos ayudara a obtener los parametros mas facil
  const { q = "" } = queryString.parse(location.search)
  const heroesMatches = useMemo(()=>getHerosByName(q as string ?? ""),[q]) 

  const {hero,handleChange,reset} = useForm({
    hero : ""
  })

  const inputRef = useRef<HTMLInputElement>(null)
  const handleSubmitSearchHero = (event:FormEvent<HTMLFormElement>) => {
    event.preventDefault()

    if(hero.trim().length) history(`?q=${hero}`)
    else history(``)
    reset()
    inputRef.current?.focus()
  }
  
  return (
    <div className="mt-2">
      <h2 className="mt-3 mb-3">Search</h2>
      <div className="row g-5">
        <div className="col-sm-12 col-md-7">
          <h4>Search Form</h4>
          <hr />
          <form className="form" onSubmit={handleSubmitSearchHero} aria-label="form">
            <input type="text" className="form-control"
            placeholder="Search Hero..."
            autoComplete="off"
            name="hero"
            value={hero}
            ref={inputRef}
            onChange={handleChange}
            />
            <button type="submit" className="btn btn-outline-primary form-control mt-3">Search ...</button>
          </form>
        </div>
        <div className="col-sm-12 col-md-5">
          <h4>Results</h4>
          <hr />
          { q == "" ? <div className="alert alert-info" aria-label="alert">Search a Hero</div> :  
            heroesMatches.length == 0 && <div className="alert alert-danger" aria-label="alert">No Results! with: { JSON.stringify(q) }</div> 
          }
          {heroesMatches.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </div>
  )
}

export default SearchPage