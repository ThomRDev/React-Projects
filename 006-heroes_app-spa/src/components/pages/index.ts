export {default as DCPage} from "./DCPage" 
export {default as LoginPage} from "./LoginPage"
export {default as MarvelPage} from "./MarvelPage"
export {default as SearchPage} from "./SearchPage"