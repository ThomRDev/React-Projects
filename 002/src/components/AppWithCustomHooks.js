
import PropTypes from "prop-types"
import useCounter from "../hooks/useCounter";
import logo from '../logo.svg';
import './../styles/App.css';

export const AppWithCustomHooks = ({ initialValue }) => {

    const { counter, actions } = useCounter(initialValue)

    const handleAction = event => {
        const { action } = event.target.dataset
        if (Object.hasOwn(actions, action)) {
            actions[action]()
        }
    }

    return (
        <div className="App">
            <header className="App-header">
                {"renderizo custom" + (new Date().getTime())}
                <img src={logo} className="App-logo" alt="logo" />
                <div className="value">{counter}</div>
                <div className='actions'>
                    <button onClick={handleAction} data-action="increment">+1</button>
                    <button onClick={() => actions["reset"]()} >Reset</button>
                    <button onClick={handleAction} data-action="decrement">-1</button>
                </div>
            </header>
        </div>
    );
}

AppWithCustomHooks.propTypes = {
    initialValue: PropTypes.number
}
AppWithCustomHooks.defaultProps = {
    initialValue: 1
}