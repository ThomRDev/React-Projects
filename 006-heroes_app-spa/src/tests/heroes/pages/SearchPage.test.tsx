import { fireEvent, getByLabelText, render, screen } from "@testing-library/react"
import { MemoryRouter, useNavigate } from "react-router-dom"
import { SearchPage } from "../../../components/pages"


const mockUseNavigate = jest.fn()

// parcial mock
jest.mock("react-router-dom",()=>({
  ...jest.requireActual("react-router-dom"),
  useNavigate : ()=>mockUseNavigate
}))

describe('Tests in SearchPage component', () => {

  beforeEach(()=>jest.clearAllMocks())

  test('should show the component correctly', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    )
    expect(container).toMatchSnapshot()

    const divAlert = getByLabelText(container,"alert") as HTMLDivElement
    expect(divAlert.innerHTML).toBe("Search a Hero")
  })
  
  test('should show to batman', () => {
    const {} = render(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchPage />
      </MemoryRouter>
    )
    expect(screen.getByText(/Batman/i)).toBeTruthy()
    const img = screen.getByRole("img")  as HTMLImageElement
    expect(img.src).toContain("/heroes/dc-batman.jpg")
    // expect(screen.getByLabelText(/alert/i)).
  })
  test("it should show the error alert if it doesn't find the hero (batman234asda)", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman234asda"]}>
        <SearchPage />
      </MemoryRouter>
    )
    const alert = screen.getByLabelText("alert") 
    expect(alert.classList).toContain("alert-danger")
  })

  // en este test tambien podria evaluar el useForm pero ya lo hice en otro test
  test('it should call the navigate hook', () => {
    render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    )

    const input = screen.getByRole("textbox") as HTMLInputElement
    fireEvent.change(input,{
      target :{
        name : "hero",
        value  :"Batman"
      },
      preventDefault(){}
    })
    expect(input.value).toBe("Batman")
    const form = screen.getByRole('form');
    fireEvent.submit( form ,{
      preventDefault(){}
    });
    // screen.debug()
    expect(input.value).toBe("")
    expect(mockUseNavigate).toHaveBeenCalled()
    expect(mockUseNavigate).toHaveBeenCalledWith(`?q=Batman`)
  })
})

// el reto es probar heropage pero todos los tests seran repetidos