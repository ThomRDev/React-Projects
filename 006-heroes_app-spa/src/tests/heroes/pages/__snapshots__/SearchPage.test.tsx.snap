// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in SearchPage component should show the component correctly 1`] = `
<div>
  <div
    class="mt-2"
  >
    <h2
      class="mt-3 mb-3"
    >
      Search
    </h2>
    <div
      class="row g-5"
    >
      <div
        class="col-sm-12 col-md-7"
      >
        <h4>
          Search Form
        </h4>
        <hr />
        <form
          aria-label="form"
          class="form"
        >
          <input
            autocomplete="off"
            class="form-control"
            name="hero"
            placeholder="Search Hero..."
            type="text"
            value=""
          />
          <button
            class="btn btn-outline-primary form-control mt-3"
            type="submit"
          >
            Search ...
          </button>
        </form>
      </div>
      <div
        class="col-sm-12 col-md-5"
      >
        <h4>
          Results
        </h4>
        <hr />
        <div
          aria-label="alert"
          class="alert alert-info"
        >
          Search a Hero
        </div>
      </div>
    </div>
  </div>
</div>
`;
